api/lib/domain/models/CertificationCandidate.js

```diff
@@ -178,19 +116,21 @@ class CertificationCandidate {
 
       set: function (complementaryCertification) {
         this.#complementaryCertification = complementaryCertification;
+        this.subscriptions = this.subscriptions.filter((subscription) => subscription.type === SubscriptionTypes.CORE);
```

Comment:
Le filter permet d'éviter d'enregistrer des doublons, des souscriptions avec un même `type`. Mais avec le recule, je me demande pourquoi on en a besoin ? Dans quel cas précis ? 🤔

Reply:
Alors ce filter récupère les subscriptions `core`.
En-dessous, s'il y a des certifs complémentaires, elles sont ajoutées au tableau de souscriptions filtrées.
Ce qui veut dire que si le candidat avait une souscription autre que `core`, elle est écrasée avec la nouvelle (pour éviter d'avoir deux inscriptions à deux complémentaire).
Ceci dit, ça vaudrait le coup de vérifier qu'il y a un usecase pour ça, ou de voir s'il n'y a pas + simple

Reply:
En fait ici on déporte la responsabilité au domain model d'être "sain". Le model, via `addSubscription` est garant pour tous les consommateurs que tu peux jamais (grace au filter) arriver dans une situation où tu permets d'avoir deux souscriptions complémentaires (ou deux coeur), c'est une règle métier.

Ce n'est ainsi pas aux X consommateurs de faire la gymnastique où ils doivent s'assurer qu'ils vont pas corrompre les règles métiers ; c'est le model qui n'est pas anémique qui assure cela.

Exemple: 
* Imaginons que tu fais pas gaffe et que le usecase de la modale fait deux fois un `addSubscription` CORE (en gros tu as un bug, ça arrive)
* Dans cette version du code actuel : aucun soucis, ton bug va pas corrompre la BDD et pas besoin d'ajouter des sécurité, tes règles métiers sont assurés
  * Rappelons d'ailleurs que le `validate` est pas toujours appelé aujourd'hui (c'est pas top, mais c'est ainsi), donc on peut pas faire confiance à Joi pour éviter une corruption
* Sans le filter, tu aurais deux items CORE dans ton array, et du coup les options suivantes pour éviter cela :
  * le repository soit il commence à ajouter des garde-fous (c'est pas bien, tu commences à mettre du métier dans tes repository)
  * tu pries pour que tes tests le voient (et bon, on a vu que nos tests sont faillibles clairement)
  * tu mets un garde-fou dans ton usecase (et du coup dans **tous** les usecases faut y penser), en gros c'est le usecase qui porte la règle métier

Pour ma part je préfère le modèle non anémique qui porte la règle métier.

Tiré du bouquin

> What’s a Domain Model?
It’s a software model of the very specific business domain you are working in. Often it’s implemented as an object model, where those objects have both data and behavior with literal and accurate business meaning.

> The Language of a team in an explicit Bounded Context expressed as a domain model adds true business value and gives us certainty that we are implementing the correct software. Even from a technical standpoint, it helps us create better models, ones with more potent behaviors, that are pure and less error prone.